---
title: "MotifStats"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{MotifStats}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```


## Introduction

`MotifStats` is a simple R package to calculate metrics to quantify the
relationship between peaks and motifs. It uses [Analysis of Motif
Enrichment (AME)](https://meme-suite.org/meme/doc/ame.html) and [Find Individual
Motif Occurrences (FIMO)](https://meme-suite.org/meme/doc/fimo.html) from the
[MEME suite](https://meme-suite.org/meme/index.html).  

<br>
The package has two distinct functions:

1. Calculate the enrichment of a given motif in a set of peaks using AME
2. Calculate the distance between each motif and its nearest peak summit. FIMO
is used to recover the locations of each motif.


## Data

The `MotifStats` package comes with motif and peak data for transcription
factors CTCF and CREB1. Details of the files are as follows:

- **CTCF TIP-seq peaks (narrowPeaks)**[^f1]
-- Retrieved from NCBI GEO under
accession
[GSE188512](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE188512)  

- **CTCF motif**
-- Retrieved from JASPAR2024 under matrix ID
[MA1930.2](http://jaspar.genereg.net/matrix/MA1930.2/)  


## Installation

`MotifStats` relies on [MEME suite](https://meme-suite.org/meme/index.html) as
a system dependency. Directions for installation can be found [here](https://www.bioconductor.org/packages/release/bioc/vignettes/memes/inst/doc/install_guide.html).  
<br>
To install the package, run the following command:  
```R
if(!require("remotes")) 
  install.packages("remotes")
remotes::install_github("neurogenomics/MotifStats")
```


## Usage

In this example analysis, we will examine the relationship between the CTCF 
motif and CTCF peaks. This includes calculating enrichment of motifs in peaks
and the distances between motifs and peak summits.


### Load packages

Load the installed package.
```{r setup_vignette}
library(MotifStats)
```

For this example, we will also load the `BSgenome.Hsapiens.UCSC.hg38` package to
provide the genome build our peaks have been derived from.
```{r include = TRUE, message = FALSE, warning = FALSE}
library(BSgenome.Hsapiens.UCSC.hg38)
```


### Prepare input data

First, we need to load the motif and peak data. The motif data obtained from
JASPAR can be loaded as follows:
```{r eval = FALSE}
name_motif <- read_motif_file(
  "/path/to/motfif/file.jaspar",
  motif_id = "id_of_motif",
  file_format = "jaspar"
  )
```

Next, we load the narrowPeak files from MACS2/3[^f2].
```{r eval = FALSE}
name_peaks <- read_peak_file("/path/to/peak/file.narrowPeak")
```

For this example, we will load the built-in CTCF TIP-seq peaks (as `GRanges`
object) and CREB motif (as `PWMatrix` object) data.
```{r include  = TRUE}
data("ctcf_motif")
data("ctcf_peaks")
```


### Calculate motif enrichment

To calculate the motif relative to a set of background sequences, we use
`motif_enrichment()`.

- Under the hood, it calls `meme::runAme` from the MEME suite. This function 
calculates the enrichment of the input motif in a set of target sequences
relative to a set of background sequences.
- A 0-order background model with the same nucleotide composition as the input
sequences is used to generate the background sequences. 
- An additional `out_dir` argument can be used to specify the
directory to save the AME output files[^f3] and the background model.  

```{r include  = TRUE}
ctcf_read_prop <- motif_enrichment(
  peak_input = ctcf_peaks,
  motif = ctcf_motif,
  genome_build = BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38,
  out_dir = "."
)
```

`ctcf_read_prop` is a list of length 3. 

- `$tp` (True positives) refers to the proportion of peaks that contain the 
motif.
- `$fp` (False positives) refers to the proportion of background sequences that
contain the motif.
- `$positive_peaks` A filtered peak set containing only those peaks that have 
the motif.

### Calculate motif-summit distances

To calculate the distance between each motif and its nearest peak summit, we use
`summit_to_motif()`.  

- `fp_rate` argument specifies the desired false-positive rate for FIMO. A
p-value is calculated using the formula:  
$$p = \frac{\text{fp_rate}}{2 * \text{promoter_length}}$$  
- A 0-order background model with the same nucleotide composition as the input
sequences is generated for comparison.  
- An additional `out_dir` argument can be used to specify the
directory to save the 0-order background file.  

```{r include  = TRUE}
ctcf_read_sum_motif <- summit_to_motif(
  peak_input = ctcf_peaks,
  motif = ctcf_motif,
  fp_rate = 0.05,
  genome_build = BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38
)
```

`ctcf_read_sum_motif` outputs a list of length 2.

- `peak_set` with peak information, as a `GRanges` object.  
- `distance_to_summit` with distances between the centre of each motif and its
nearest peak summit.  

**NOTE**: When a motif is found multiple times within a single peak, the
`peak_set` objects will contain multiple entries (rows) corresponding to the 
same peak. Each of these entries represents a distinct occurrence of the motif 
within that peak.

### Visualize results

We can optionally visualise the distribution of distances by using
`density_plot()`.
```{r include  = TRUE, fig.width = 7, fig.height = 4}
density_plot(
  ctcf_read_sum_motif$distance_to_summit,
  plot_title = "CTCF motif distance to summit",
  x_label = "Distance to summit (bp)",
  y_label = "Density"
)
```

Notice how the distribution of summit-to-motif distances is centred on 0. This
suggests that the peak summits are correctly profiling transcription factor 
binding sites.


> **NOTE:** Since AME and FIMO accept different parameters and are calculated
independently, it is not possible to obtain directly comparable results.


## Session Info

<details>

```{r echo = FALSE}
sessionInfo()
```

</details>


<!-- Footnotes -->
[^f1]: The peak file is a subset of chromosome 19 to reduce the file size.
[^f2]: narrowPeak files from both version 2 and 3 of [MACS: Model-based Analysis
for ChIP-Seq](https://github.com/macs3-project/MACS) is supported.
[^f3]: For more information on the output files, refer to the
[AME documentation](https://meme-suite.org/meme/doc/ame.html).
